x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/vectors/TextEmbeddingQueryVectorBuilder.java

```diff
@@ -0,0 +1,141 @@
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+import org.elasticsearch.xpack.core.ml.action.InferModelAction;
+import org.elasticsearch.xpack.core.ml.inference.results.TextEmbeddingResults;
+import org.elasticsearch.xpack.core.ml.inference.results.WarningInferenceResults;
+import org.elasticsearch.xpack.core.ml.inference.trainedmodel.TextEmbeddingConfigUpdate;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+import static org.elasticsearch.xcontent.ConstructingObjectParser.constructorArg;
+
+public class TextEmbeddingQueryVectorBuilder implements QueryVectorBuilder {
+
+    public static final String NAME = "text_embedding";
+
+    public static final ParseField MODEL_TEXT = new ParseField("model_text");
+
+    public static final ConstructingObjectParser<TextEmbeddingQueryVectorBuilder, Void> PARSER = new ConstructingObjectParser<>(
+        NAME,
+        args -> new TextEmbeddingQueryVectorBuilder((String) args[0], (String) args[1])
+    );
+
+    static {
+        PARSER.declareString(constructorArg(), InferModelAction.Request.MODEL_ID);
+        PARSER.declareString(constructorArg(), MODEL_TEXT);
+    }
+
+    public static TextEmbeddingQueryVectorBuilder fromXContent(XContentParser parser) throws IOException {
+        return PARSER.parse(parser, null);
+    }
+
+    private final String modelId;
+    private final String modelText;
+
+    public TextEmbeddingQueryVectorBuilder(String modelId, String modelText) {
+        this.modelId = modelId;
+        this.modelText = modelText;
+    }
+
+    public TextEmbeddingQueryVectorBuilder(StreamInput in) throws IOException {
+        this.modelId = in.readString();
+        this.modelText = in.readString();
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersion.V_8_7_0;
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(modelId);
+        out.writeString(modelText);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(InferModelAction.Request.MODEL_ID.getPreferredName(), modelId);
+        builder.field(MODEL_TEXT.getPreferredName(), modelText);
+        builder.endObject();
+        return builder;
+    }
+
+    @Override
+    public void buildVector(Client client, ActionListener<float[]> listener) {
+        InferModelAction.Request inferRequest = InferModelAction.Request.forTextInput(
+            modelId,
+            TextEmbeddingConfigUpdate.EMPTY_INSTANCE,
+            List.of(modelText)
+        );
+
+        client.execute(InferModelAction.INSTANCE, inferRequest, ActionListener.wrap(response -> {
```

Comment:
Which user is this running as? I think we agreed in a meeting last year that we wouldn't require any extra permissions for semantic search over and above the ability to search the index being searched. So this needs to be running using the internal `_xpack` user.